internal/pkg/summary/mustgather.go

```diff
@@ -0,0 +1,319 @@
+}
+
+func (mg *MustGather) Process(buf *bytes.Buffer) error {
+	log.Debugf("Processing results/Populating/Populating Summary/Processing/MustGather/Reading")
+	tar, err := mg.read(buf)
+	if err != nil {
+		return err
+	}
+	log.Debugf("Processing results/Populating/Populating Summary/Processing/MustGather/Processing")
+	err = mg.extract(tar)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (mg *MustGather) read(buf *bytes.Buffer) (*tar.Reader, error) {
+	file, err := xz.NewReader(buf)
+	if err != nil {
+		log.Fatalf("NewReader error %s", err)
+	}
+	return tar.NewReader(file), nil
+}
+
+func (mg *MustGather) matchToExtract(path string) (bool, string) {
+	patterns := make(map[string]string, 1)
+	patterns["logs"] = `(\/namespaces\/.*\/pods\/.*.log)`
+	patterns["events"] = `(\/event-filter.html)`
+	// TODO /host_service_logs/.*.log
+	for typ, pattern := range patterns {
+		re := regexp.MustCompile(pattern)
+		if re.MatchString(path) {
+			return true, typ
+		}
+	}
+	return false, ""
+}
+
+// extractRelativePath removes the prefix of must-gather path/image to save the
+// relative file path when extracting the file or mapping in the counters.
+// OPCT collects must-gather automatically saving in the directory must-gather-opct.
+func (mg *MustGather) extractRelativePath(file string) string {
+	re := regexp.MustCompile(`must-gather-opct/([A-Za-z0-9]+(-[A-Za-z0-9]+)+\/)`)
+
+	split := re.Split(file, -1)
+	if len(split) != 2 {
+		return file
+	}
+	return split[1]
+}
+
+// extract dispatch to process must-gather items.
+func (mg *MustGather) extract(tarball *tar.Reader) error {
+
+	// Create must-gather directory
+	if _, err := os.Stat(mg.path); err != nil {
+		if err := os.MkdirAll(mg.path, 0755); err != nil {
+			return err
+		}
+	}
+
+	// Leaky bucket implementation (queue limit) to parallel process must-gather items
+	// without exhausting resources.
+	// Benckmark info: this parallel processing decreased 3 times the total processing time.
+	// Samples: Serial=~100s, rate(100)=~30s, rate(150)=~25s.
+	waiterProcNS := &sync.WaitGroup{}
+	chProcNSErrors := make(chan *MustGatherLog, 30)
+	semaphore := make(chan struct{}, 50)
+	// have a max rate of N/sec
+	rate := make(chan struct{}, 100)
+	for i := 0; i < cap(rate); i++ {
+		rate <- struct{}{}
+	}
+	// leaky bucket
+	go func() {
+		ticker := time.NewTicker(100 * time.Millisecond)
+		defer ticker.Stop()
+		for range ticker.C {
+			_, ok := <-rate
+			// if this isn't going to run indefinitely, signal
+			// this to return by closing the rate channel.
+			if !ok {
+				return
+			}
+		}
+	}()
+	// consumer
+	go func() {
+		for mgLog := range chProcNSErrors {
+			err := mg.processNamespaceErrors(mgLog)
+			if err != nil {
+				log.Errorf("error extracting log info: %v", err)
+			}
+			waiterProcNS.Done()
+		}
+	}()
+
+	// Walk through files in must-gather tarball file.
+	for {
+		header, err := tarball.Next()
+
+		switch {
+		// no more files
+		case err == io.EOF:
+			return nil
```

Comment:
maybe we need to exit the loop if we want to return to the regular function execution.